node/src/components/binary_port.rs

```diff
@@ -0,0 +1,887 @@
+        + From<AcceptTransactionRequest>
+        + From<NetworkInfoRequest>
+        + From<ReactorInfoRequest>
+        + From<ConsensusRequest>
+        + From<BlockSynchronizerRequest>
+        + From<UpgradeWatcherRequest>
+        + From<ChainspecRawBytesRequest>
+        + Send,
+{
+    let version = effect_builder.get_protocol_version().await;
+    match req {
+        BinaryRequest::TryAcceptTransaction { transaction } => {
+            metrics.binary_port_try_accept_transaction_count.inc();
+            try_accept_transaction(effect_builder, transaction, None, version).await
+        }
+        BinaryRequest::TrySpeculativeExec {
+            transaction,
+            state_root_hash,
+            block_time,
+            protocol_version,
+            speculative_exec_at_block,
+        } => {
+            metrics.binary_port_try_speculative_exec_count.inc();
+            let response = try_accept_transaction(
+                effect_builder,
+                transaction.clone(),
+                Some(speculative_exec_at_block),
+                version,
+            )
+            .await;
+            if !response.is_success() {
+                return response;
+            }
+            try_speculative_execution(
+                effect_builder,
+                state_root_hash,
+                block_time,
+                protocol_version,
+                transaction,
+            )
+            .await
+        }
+        BinaryRequest::Get(get_req) => {
+            handle_get_request(get_req, effect_builder, config, metrics, version).await
+        }
+    }
+}
+
+async fn handle_get_request<REv>(
+    get_req: GetRequest,
+    effect_builder: EffectBuilder<REv>,
+    config: &Config,
+    metrics: &Metrics,
+    protocol_version: ProtocolVersion,
+) -> BinaryResponse
+where
+    REv: From<Event>
+        + From<StorageRequest>
+        + From<NetworkInfoRequest>
+        + From<ReactorInfoRequest>
+        + From<ConsensusRequest>
+        + From<BlockSynchronizerRequest>
+        + From<UpgradeWatcherRequest>
+        + From<ChainspecRawBytesRequest>
+        + From<ContractRuntimeRequest>
+        + Send,
+{
+    match get_req {
+        // this workaround is in place because get_block_transfers performs a lazy migration
+        GetRequest::Db { db_tag, key } if db_tag == u8::from(DbId::Transfer) => {
+            metrics.binary_port_get_db_count.inc();
+            let Ok(block_hash) = bytesrepr::deserialize_from_slice(&key) else {
+                return BinaryResponse::new_error(binary_port::ErrorCode::BadRequest, protocol_version);
+            };
+            let Some(transfers) = effect_builder
+                .get_block_transfers_from_storage(block_hash)
+                .await else {
+                return BinaryResponse::new_empty(protocol_version);
+            };
+            let serialized =
+                bincode::serialize(&transfers).expect("should serialize transfers to bytes");
```

Comment:
might need to refactor the code here a bit to do that, I'll leave it for next week

Reply:
yes, definitely.
good catch!
I did my usual "unwrap()/expect()" internal review pass on the code but I must have missed some of them.

Reply:
I've updated this to return an InternalError now, although we might also want to add a debug log or something because the InternalError isnt particularly meaningful